File: SDKs/flutter/lib/src/ivx_ai_client.dart
import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

import 'types.dart';

// ---------------------------------------------------------------------------
// Models
// ---------------------------------------------------------------------------

/// Represents a response when starting an AI voice or host session.
class IVXAISessionResponse {
  final String sessionId;
  final String status;
  final Map<String, dynamic> metadata;

  const IVXAISessionResponse({
    required this.sessionId,
    this.status = 'active',
    this.metadata = const {},
  });

  factory IVXAISessionResponse.fromJson(Map<String, dynamic> json) =>
      IVXAISessionResponse(
        sessionId: json['session_id'] as String? ?? '',
        status: json['status'] as String? ?? 'active',
        metadata: json['metadata'] as Map<String, dynamic>? ?? const {},
      );

  @override
  String toString() =>
      'IVXAISessionResponse(sessionId: $sessionId, status: $status)';
}

/// A single message within an AI conversation.
class IVXAIMessage {
  final String id;
  final String role;
  final String content;
  final DateTime timestamp;
  final Map<String, dynamic> metadata;

  const IVXAIMessage({
    required this.id,
    required this.role,
    required this.content,
    required this.timestamp,
    this.metadata = const {},
  });

  factory IVXAIMessage.fromJson(Map<String, dynamic> json) => IVXAIMessage(
        id: json['id'] as String? ?? '',
        role: json['role'] as String? ?? 'assistant',
        content: json['content'] as String? ?? '',
        timestamp: DateTime.tryParse(json['timestamp'] as String? ?? '') ??
            DateTime.now(),
        metadata: json['metadata'] as Map<String, dynamic>? ?? const {},
      );

  @override
  String toString() => 'IVXAIMessage(id: $id, role: $role)';
}

/// Profile data supplied when creating an AI-hosted session.
class IVXHostProfile {
  final String name;
  final String personaId;
  final String style;
  final Map<String, dynamic> config;

  const IVXHostProfile({
    required this.name,
    required this.personaId,
    this.style = 'default',
    this.config = const {},
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        'persona_id': personaId,
        'style': style,
        'config': config,
      };

  @override
  String toString() => 'IVXHostProfile(name: $name, personaId: $personaId)';
}

/// Entitlement status for AI features.
class IVXAIEntitlement {
  final bool hasAccess;
  final String tier;
  final int remainingCredits;
  final DateTime? expiresAt;

  const IVXAIEntitlement({
    required this.hasAccess,
    this.tier = 'free',
    this.remainingCredits = 0,
    this.expiresAt,
  });

  factory IVXAIEntitlement.fromJson(Map<String, dynamic> json) =>
      IVXAIEntitlement(
        hasAccess: json['has_access'] as bool? ?? false,
        tier: json['tier'] as String? ?? 'free',
        remainingCredits: json['remaining_credits'] as int? ?? 0,
        expiresAt: json['expires_at'] != null
            ? DateTime.tryParse(json['expires_at'] as String)
            : null,
      );

  @override
  String toString() =>
      'IVXAIEntitlement(tier: $tier, hasAccess: $hasAccess, credits: $remainingCredits)';
}

/// Descriptor for an available AI persona.
class IVXAIPersona {
  final String id;
  final String name;
  final String description;
  final String avatarUrl;
  final List<String> supportedLanguages;

  const IVXAIPersona({
    required this.id,
    required this.name,
    this.description = '',
    this.avatarUrl = '',
    this.supportedLanguages = const ['en'],
  });

  factory IVXAIPersona.fromJson(Map<String, dynamic> json) => IVXAIPersona(
        id: json['id'] as String? ?? '',
        name: json['name'] as String? ?? '',
        description: json['description'] as String? ?? '',
        avatarUrl: json['avatar_url'] as String? ?? '',
        supportedLanguages:
            (json['supported_languages'] as List<dynamic>?)
                    ?.cast<String>() ??
                const ['en'],
      );

  @override
  String toString() => 'IVXAIPersona(id: $id, name: $name)';
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

/// HTTP client for the IntelliVerseX AI service.
///
/// Provides voice-session management, text messaging, AI-hosted game
/// sessions, entitlement checks, and persona queries against a remote API.
///
/// ```dart
/// final ai = IVXAIClient.instance;
/// await ai.initialize(apiBaseUrl: 'https://ai.example.com');
/// final session = await ai.startVoiceSession('persona_1', userId);
/// ```
class IVXAIClient {
  static IVXAIClient? _instance;

  late String _baseUrl;
  String? _apiKey;
  bool _initialized = false;
  late http.Client _httpClient;

  IVXAIClient._();

  /// Singleton accessor.
  static IVXAIClient get instance => _instance ??= IVXAIClient._();

  /// Reset the singleton (useful for testing).
  static void resetInstance() => _instance = null;

  /// Whether [initialize] has been called successfully.
  bool get isInitialized => _initialized;

  // ---------------------------------------------------------------------------
  // Initialization
  // ---------------------------------------------------------------------------

  /// Set up the AI client with the remote [apiBaseUrl].
  ///
  /// An optional [apiKey] is sent as a `Bearer` token on every request.
  Future<void> initialize({
    required String apiBaseUrl,
    String? apiKey,
  }) async {
    if (apiBaseUrl.trim().isEmpty) {
      throw const IVXError(code: -1, message: 'apiBaseUrl cannot be empty.');
    }
    _baseUrl = apiBaseUrl.endsWith('/') ? apiBaseUrl : '$apiBaseUrl/';
    _apiKey = apiKey;
    _httpClient = http.Client();
    _initialized = true;
    _log('AI client initialized — $_baseUrl');
  }

  // ---------------------------------------------------------------------------
  // Voice sessions
  // ---------------------------------------------------------------------------

  /// Start a voice conversation session for [personaId] on behalf of [userId].
  ///
  /// An optional [language] tag (e.g. `"en"`, `"es"`) sets the session locale.
  Future<IVXAISessionResponse> startVoiceSession(
    String personaId,
    String userId, {
    String? language,
  }) async {
    _ensureInitialized();
    final body = {
      'persona_id': personaId,
      'user_id': userId,
      if (language != null) 'language': language,
    };
    final json = await _post('v1/voice/sessions', body);
    return IVXAISessionResponse.fromJson(json);
  }

  /// Gracefully end the voice session identified by [sessionId].
  Future<void> endVoiceSession(String sessionId) async {
    _ensureInitialized();
    await _post('v1/voice/sessions/$sessionId/end', {});
    _log('Voice session $sessionId ended');
  }

  // ---------------------------------------------------------------------------
  // Text messaging
  // ---------------------------------------------------------------------------

  /// Send a [text] message into an active session.
  Future<void> sendText(String sessionId, String text) async {
    _ensureInitialized();
    await _post('v1/sessions/$sessionId/messages', {'text': text});
  }

  /// Poll new messages for [sessionId], optionally filtering to those after
  /// [since].
  Future<List<IVXAIMessage>> pollMessages(
    String sessionId, {
    DateTime? since,
  }) async {
    _ensureInitialized();
    final query = <String, String>{};
    if (since != null) query['since'] = since.toUtc().toIso8601String();
    final json = await _get('v1/sessions/$sessionId/messages', query);
    final items = json['messages'] as List<dynamic>? ?? [];
    return items
        .cast<Map<String, dynamic>>()
        .map(IVXAIMessage.fromJson)
        .toList();
  }

  // ---------------------------------------------------------------------------
  // Host sessions
  // ---------------------------------------------------------------------------

  /// Create an AI-hosted game session tied to [matchId].
  Future<IVXAISessionResponse> startHostSession(
    String matchId,
    IVXHostProfile profile,
  ) async {
    _ensureInitialized();
    final body = {
      'match_id': matchId,
      ...profile.toJson(),
    };
    final json = await _post('v1/host/sessions', body);
    return IVXAISessionResponse.fromJson(json);
  }

  /// Push a game event of [eventType] with JSON [data] into a host session.
  Future<void> sendHostEvent(
    String sessionId,
    String eventType,
    String data,
  ) async {
    _ensureInitialized();
    await _post('v1/host/sessions/$sessionId/events', {
      'event_type': eventType,
      'data': data,
    });
  }

  // ---------------------------------------------------------------------------
  // Entitlements & personas
  // ---------------------------------------------------------------------------

  /// Check AI feature entitlement for [userId].
  Future<IVXAIEntitlement> checkEntitlement(String userId) async {
    _ensureInitialized();
    final json = await _get('v1/entitlements/$userId', {});
    return IVXAIEntitlement.fromJson(json);
  }

  /// Retrieve the list of available AI personas.
  Future<List<IVXAIPersona>> getPersonas() async {
    _ensureInitialized();
    final json = await _get('v1/personas', {});
    final items = json['personas'] as List<dynamic>? ?? [];
    return items
        .cast<Map<String, dynamic>>()
        .map(IVXAIPersona.fromJson)
        .toList();
  }

  // ---------------------------------------------------------------------------
  // HTTP helpers
  // ---------------------------------------------------------------------------

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        if (_apiKey != null) 'Authorization': 'Bearer $_apiKey',
      };

  Future<Map<String, dynamic>> _post(
    String path,
    Map<String, dynamic> body,
  ) async {
    final uri = Uri.parse('$_baseUrl$path');
    final response = await _httpClient.post(
      uri,
      headers: _headers,
      body: jsonEncode(body),
    );
    return _handleResponse(response);
  }

  Future<Map<String, dynamic>> _get(
    String path,
    Map<String, String> queryParams,
  ) async {
    final uri =
        Uri.parse('$_baseUrl$path').replace(queryParameters: queryParams);
    final response = await _httpClient.get(uri, headers: _headers);
    return _handleResponse(response);
  }

  Map<String, dynamic> _handleResponse(http.Response response) {
    if (response.statusCode >= 200 && response.statusCode < 300) {
      if (response.body.isEmpty) return {};
      final decoded = jsonDecode(response.body);
      if (decoded is Map<String, dynamic>) return decoded;
      return {'data': decoded};
    }
    throw IVXError(
      code: response.statusCode,
      message: 'AI API error ${response.statusCode}: ${response.body}',
    );
  }

  // ---------------------------------------------------------------------------
  // Internal
  // ---------------------------------------------------------------------------

  void _ensureInitialized() {
    if (!_initialized) {
      throw const IVXError(
        code: -1,
        message: 'IVXAIClient not initialized. Call initialize() first.',
      );
    }
  }

  void _log(String message) {
    // ignore: avoid_print
    print('[IntelliVerseX:AI] $message');
  }
}

File: SDKs/cocos2dx/Classes/IntelliVerseX/IVXAIClient.cpp
// Copyright (c) 2026 Intelli-verse-X
// MIT License — see LICENSE in the project root.

#include "IntelliVerseX/IVXAIClient.h"
#include "cocos2d.h"
#include "network/HttpClient.h"
#include "network/HttpRequest.h"
#include "network/HttpResponse.h"
#include "json/rapidjson.h"
#include "json/document.h"
#include "json/writer.h"
#include "json/stringbuffer.h"

namespace IntelliVerseX {

IVXAIClient& IVXAIClient::getInstance() {
    static IVXAIClient instance;
    return instance;
}

void IVXAIClient::initialize(const std::string& apiBaseUrl,
                              const std::string& apiKey,
                              bool enableDebugLogs) {
    _apiBaseUrl = apiBaseUrl;
    _apiKey = apiKey;
    _enableDebugLogs = enableDebugLogs;
    _initialized = true;
    log("AI client initialized — " + apiBaseUrl);
}

void IVXAIClient::startVoiceSession(const std::string& personaId,
                                     const std::string& userId,
                                     AISessionCallback onSuccess,
                                     ErrorCallback onError) {
    if (!_initialized) {
        if (onError) onError({-1, "AI client not initialized"});
        return;
    }

    rapidjson::StringBuffer sb;
    rapidjson::Writer<rapidjson::StringBuffer> w(sb);
    w.StartObject();
    w.Key("personaId"); w.String(personaId.c_str());
    w.Key("userId"); w.String(userId.c_str());
    w.EndObject();

    httpPost("/ai-voice/session", sb.GetString(),
        [this, onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            if (doc.HasParseError() || !doc.IsObject()) {
                return;
            }
            AISessionResponse session;
            session.sessionId = doc.HasMember("sessionId") ? doc["sessionId"].GetString() : "";
            session.personaId = doc.HasMember("personaId") ? doc["personaId"].GetString() : "";
            session.userId = doc.HasMember("userId") ? doc["userId"].GetString() : "";
            session.status = doc.HasMember("status") ? doc["status"].GetString() : "";
            session.createdAt = doc.HasMember("createdAt") ? doc["createdAt"].GetInt64() : 0;
            log("Voice session started: " + session.sessionId);
            if (onSuccess) onSuccess(session);
        }, onError);
}

void IVXAIClient::endVoiceSession(const std::string& sessionId,
                                   SuccessCallback onSuccess,
                                   ErrorCallback onError) {
    if (!_initialized) {
        if (onError) onError({-1, "AI client not initialized"});
        return;
    }

    httpPost("/ai-voice/session/" + sessionId + "/end", "{}",
        [this, sessionId, onSuccess](const std::string&) {
            log("Voice session ended: " + sessionId);
            if (onSuccess) onSuccess();
        }, onError);
}

void IVXAIClient::sendText(const std::string& sessionId,
                            const std::string& text,
                            AIMessageCallback onSuccess,
                            ErrorCallback onError) {
    if (!_initialized) {
        if (onError) onError({-1, "AI client not initialized"});
        return;
    }

    rapidjson::StringBuffer sb;
    rapidjson::Writer<rapidjson::StringBuffer> w(sb);
    w.StartObject();
    w.Key("text"); w.String(text.c_str());
    w.EndObject();

    httpPost("/ai-voice/session/" + sessionId + "/text", sb.GetString(),
        [onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            if (doc.HasParseError() || !doc.IsObject()) {
                return;
            }
            AIMessage msg;
            msg.id = doc.HasMember("id") ? doc["id"].GetString() : "";
            msg.sessionId = doc.HasMember("sessionId") ? doc["sessionId"].GetString() : "";
            msg.role = doc.HasMember("role") ? doc["role"].GetString() : "";
            msg.text = doc.HasMember("text") ? doc["text"].GetString() : "";
            msg.timestamp = doc.HasMember("timestamp") ? doc["timestamp"].GetInt64() : 0;
            if (onSuccess) onSuccess(msg);
        }, onError);
}

void IVXAIClient::startHostSession(const std::string& matchId,
                                    const HostProfile& profile,
                                    AISessionCallback onSuccess,
                                    ErrorCallback onError) {
    if (!_initialized) {
        if (onError) onError({-1, "AI client not initialized"});
        return;
    }

    rapidjson::StringBuffer sb;
    rapidjson::Writer<rapidjson::StringBuffer> w(sb);
    w.StartObject();
    w.Key("matchId"); w.String(matchId.c_str());
    w.Key("profile");
    w.StartObject();
    w.Key("displayName"); w.String(profile.displayName.c_str());
    if (!profile.avatarUrl.empty()) {
        w.Key("avatarUrl"); w.String(profile.avatarUrl.c_str());
    }
    if (!profile.metadata.empty()) {
        w.Key("metadata"); w.String(profile.metadata.c_str());
    }
    w.EndObject();
    w.EndObject();

    httpPost("/ai-host/session", sb.GetString(),
        [this, onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            if (doc.HasParseError() || !doc.IsObject()) {
                return;
            }
            AISessionResponse session;
            session.sessionId = doc.HasMember("sessionId") ? doc["sessionId"].GetString() : "";
            session.personaId = doc.HasMember("personaId") ? doc["personaId"].GetString() : "";
            session.userId = doc.HasMember("userId") ? doc["userId"].GetString() : "";
            session.status = doc.HasMember("status") ? doc["status"].GetString() : "";
            session.createdAt = doc.HasMember("createdAt") ? doc["createdAt"].GetInt64() : 0;
            log("Host session started: " + session.sessionId);
            if (onSuccess) onSuccess(session);
        }, onError);
}

void IVXAIClient::sendHostEvent(const std::string& sessionId,
                                 const std::string& eventType,
                                 const std::string& data,
                                 SuccessCallback onSuccess,
                                 ErrorCallback onError) {
    if (!_initialized) {
        if (onError) onError({-1, "AI client not initialized"});
        return;
    }

    rapidjson::StringBuffer sb;
    rapidjson::Writer<rapidjson::StringBuffer> w(sb);
    w.StartObject();
    w.Key("eventType"); w.String(eventType.c_str());
    w.Key("data"); w.String(data.c_str());
    w.EndObject();

    httpPost("/ai-host/session/" + sessionId + "/event", sb.GetString(),
        [this, eventType, onSuccess](const std::string&) {
            log("Host event sent: " + eventType);
            if (onSuccess) onSuccess();
        }, onError);
}

void IVXAIClient::checkEntitlement(const std::string& userId,
                                    AIEntitlementCallback onSuccess,
                                    ErrorCallback onError) {
    if (!_initialized) {
        if (onError) onError({-1, "AI client not initialized"});
        return;
    }

    httpGet("/ai-voice/entitlement/" + userId,
        [onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            if (doc.HasParseError() || !doc.IsObject()) {
                return;
            }
            AIEntitlement ent;
            ent.userId = doc.HasMember("userId") ? doc["userId"].GetString() : "";
            ent.entitled = doc.HasMember("entitled") ? doc["entitled"].GetBool() : false;
            ent.remainingCredits = doc.HasMember("remainingCredits") ? doc["remainingCredits"].GetInt() : 0;
            ent.plan = doc.HasMember("plan") ? doc["plan"].GetString() : "";
            if (onSuccess) onSuccess(ent);
        }, onError);
}

void IVXAIClient::getPersonas(AIPersonasCallback onSuccess,
                               ErrorCallback onError) {
    if (!_initialized) {
        if (onError) onError({-1, "AI client not initialized"});
        return;
    }

    httpGet("/ai-voice/personas",
        [onSuccess](const std::string& response) {
            rapidjson::Document doc;
            doc.Parse(response.c_str());
            std::vector<AIPersona> personas;
            if (!doc.HasParseError() && doc.IsArray()) {
                for (rapidjson::SizeType i = 0; i < doc.Size(); ++i) {
                    const auto& obj = doc[i];
                    if (!obj.IsObject()) continue;
                    AIPersona p;
                    p.id = obj.HasMember("id") ? obj["id"].GetString() : "";
                    p.name = obj.HasMember("name") ? obj["name"].GetString() : "";
                    p.description = obj.HasMember("description") ? obj["description"].GetString() : "";
                    p.avatarUrl = obj.HasMember("avatarUrl") ? obj["avatarUrl"].GetString() : "";
                    if (obj.HasMember("supportedLanguages") && obj["supportedLanguages"].IsArray()) {
                        for (rapidjson::SizeType j = 0; j < obj["supportedLanguages"].Size(); ++j) {
                            p.supportedLanguages.push_back(obj["supportedLanguages"][j].GetString());
                        }
                    }
                    personas.push_back(p);
                }
            }
            if (onSuccess) onSuccess(personas);
        }, onError);
}

// ---------------------------------------------------------------------------
// HTTP helpers (cocos2d::network::HttpClient)
// ---------------------------------------------------------------------------

void IVXAIClient::httpPost(const std::string& path,
                            const std::string& bodyJson,
                            std::function<void(const std::string&)> onSuccess,
                            ErrorCallback onError) {
    auto request = new (std::nothrow) cocos2d::network::HttpRequest();
    if (!request) return;

    request->setRequestType(cocos2d::network::HttpRequest::Type::POST);
    request->setUrl((_apiBaseUrl + path).c_str());
    request->setRequestData(bodyJson.c_str(), bodyJson.size());

    std::vector<std::string> headers;
    headers.push_back("Content-Type: application/json");
    headers.push_back("Authorization: Bearer " + _apiKey);
    request->setHeaders(headers);

    request->setResponseCallback([this, onSuccess, onError](
            cocos2d::network::HttpClient* /*client*/,
            cocos2d::network::HttpResponse* resp) {
        if (!resp || !resp->isSucceed()) {
            std::string errMsg = resp ? resp->getErrorBuffer() : "Request failed";
            long statusCode = resp ? resp->getResponseCode() : -1;
            log("HTTP error: " + errMsg);
            if (onError) onError({static_cast<int>(statusCode), errMsg});
            return;
        }
        auto* data = resp->getResponseData();
        std::string body(data->begin(), data->end());
        if (onSuccess) onSuccess(body);
    });

    cocos2d::network::HttpClient::getInstance()->send(request);
    request->release();
}

void IVXAIClient::httpGet(const std::string& path,
                           std::function<void(const std::string&)> onSuccess,
                           ErrorCallback onError) {
    auto request = new (std::nothrow) cocos2d::network::HttpRequest();
    if (!request) return;

    request->setRequestType(cocos2d::network::HttpRequest::Type::GET);
    request->setUrl((_apiBaseUrl + path).c_str());

    std::vector<std::string> headers;
    headers.push_back("Content-Type: application/json");
    headers.push_back("Authorization: Bearer " + _apiKey);
    request->setHeaders(headers);

    request->setResponseCallback([this, onSuccess, onError](
            cocos2d::network::HttpClient* /*client*/,
            cocos2d::network::HttpResponse* resp) {
        if (!resp || !resp->isSucceed()) {
            std::string errMsg = resp ? resp->getErrorBuffer() : "Request failed";
            long statusCode = resp ? resp->getResponseCode() : -1;
            log("HTTP error: " + errMsg);
            if (onError) onError({static_cast<int>(statusCode), errMsg});
            return;
        }
        auto* data = resp->getResponseData();
        std::string body(data->begin(), data->end());
        if (onSuccess) onSuccess(body);
    });

    cocos2d::network::HttpClient::getInstance()->send(request);
    request->release();
}

void IVXAIClient::log(const std::string& message) {
    if (_enableDebugLogs) {
        cocos2d::log("[IntelliVerseX:AI] %s", message.c_str());
    }
}

} // namespace IntelliVerseX
